staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go

```diff
@@ -55,9 +55,44 @@ var (
 	DefaultFeatureGate FeatureGate = NewFeatureGate()
 )
 
+// FeatureSpec describes a feature that can be used by many components
 type FeatureSpec struct {
-	Default    bool
+	// Default enables the feature by default
+	Default bool
+	// PreRelease set the release state of the feature
 	PreRelease prerelease
+	// Components list the components that is using the feature
+	Components []string
```

Comment:
We have got many questions related to this mapping. Many many thanks for figuring this out.

One improvement I'd like to suggest is to introduce 3 additional string fields:

```
  // IntroducedSince is the release when a feature is first introduced
  IntroducedSince: string
  // DeprecatedSince is the release when the feature is marked as deprecated
  DeprecatedSince: string
  // RemovedSince is the release when a feature is removed
  RemovedSince: string
```

The `RemovedSince` field can be ignored because it disappears anyway, however the other two fields would be very helpful for users.

Reply:
Sounds like a good idea. However, we have `PreRelease: utilfeature.Alpha, utilfeature.Beta, utilfeature.GA`, how can these versions be mapping to your suggested fields?

Reply:
I like this idea too. I think `IntroducedSince` could also be a `Version` struct, since that provides more semantic context than a string (which can be malformed). IMO the `PreRelease` is something slightly orthogonal to what version a feature was added.

Reply:
Maybe this should be another change ? As of now i dont think the added metadata ( that replaces/augments current versioning metadata in feature comments ) makes much sense with the current implementation since they shouldnt appear in flag text.

This probably makes more sense if we autogenerate some form of documentation similar to the matrix in https://github.com/kubernetes/website/issues/6141#issuecomment-341426400 for example.

Reply:
Adding these fields would serve more than just docs, it would allow us to, for example, validate that the feature can be used with a specific component version. But yeah, we can do this in a follow-up PR if you'd prefer.

Reply:
>  it would allow us to, for example, validate that the feature can be used with a specific component version

Indeed, but it probably should be done carefully. What should we do if the validation failed ? We can prevent the component from starting but that will probably break some installations if people adding their required flags to all components ( that's what i have been doing before this change :stuck_out_tongue_closed_eyes: )

Reply:
Yeah, good points. Would you mind creating an issue so we can track this feature conversation?